safehttp/form.go

```diff
@@ -0,0 +1,219 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package safehttp
+
+import (
+	"errors"
+	"mime/multipart"
+	"strconv"
+)
+
+// Form TODO
+type Form struct {
+	values map[string][]string
+	err    error
+	parsed bool
+}
+
+// MultipartForm TODO
+type MultipartForm struct {
+	form Form
+	file map[string][]*multipart.FileHeader
+}
+
+// Int TODO
+func (f *Form) Int(paramName string, defaultValue int) int {
+	if f.err != nil {
+		return defaultValue
+	}
+	if !f.parsed {
+		panic("form has not been parsed")
+	}
+	vals, ok := f.values[paramName]
+	if !ok {
+		f.err = errors.New("no value found for key " + paramName)
+		return defaultValue
+	}
+	paramVal, err := strconv.Atoi(vals[0])
+	if err != nil {
+		f.err = err
+		return defaultValue
+	}
+	return paramVal
+}
+
+// Uint TODO
+func (f *Form) Uint(paramName string, defaultValue uint64) uint64 {
+	if f.err != nil {
+		return defaultValue
+	}
+	if !f.parsed {
+		panic("form has not been parsed")
+	}
+	vals, ok := f.values[paramName]
+	if !ok {
+		f.err = errors.New("no value found for key " + paramName)
+		return defaultValue
+	}
+	paramVal, err := strconv.ParseUint(vals[0], 10, 0)
+	if err != nil {
+		f.err = err
+		return defaultValue
+	}
+	return paramVal
+}
+func (f *Form) String(paramName string, defaultValue string) string {
+	if f.err != nil {
+		return defaultValue
+	}
+	if !f.parsed {
+		panic("form has not been parsed")
+	}
+	vals, ok := f.values[paramName]
+	if !ok {
+		f.err = errors.New("no value found for key " + paramName)
+		return defaultValue
+	}
+	return vals[0]
+}
+
+// Float64 TODO
+func (f *Form) Float64(paramName string, defaultValue float64) float64 {
+	if f.err != nil {
+		return defaultValue
+	}
+	if !f.parsed {
+		panic("form has not been parsed")
+	}
+	vals, ok := f.values[paramName]
+	if !ok {
+		f.err = errors.New("no value found for key " + paramName)
+		return defaultValue
+	}
+	paramVal, err := strconv.ParseFloat(vals[0], 64)
+	if err != nil {
+		f.err = err
+		return defaultValue
+	}
+	return paramVal
+}
+
+// Bool TODO
+func (f *Form) Bool(paramName string, defaultValue bool) bool {
+	if f.err != nil {
+		return defaultValue
+	}
+	if !f.parsed {
+		panic("form has not been parsed")
+	}
+	vals, ok := f.values[paramName]
+	if !ok {
+		f.err = errors.New("no value found for key " + paramName)
+		return defaultValue
+	}
+	if vals[0] != "true" {
+		if vals[0] != "false" {
+			f.err = errors.New("values of form parameter " + paramName + " not a boolean")
+		}
+		return false
+	}
+	return true
+}
+
+// Slice TODO
+func (f *Form) Slice(slice interface{}, paramName string) {
+	if f.err != nil {
+		slice = nil
+		return
+	}
+	if !f.parsed {
+		panic("form has not been parsed")
+	}
+	vals, ok := f.values[paramName]
+	if !ok {
+		f.err = errors.New("no value found for key " + paramName)
+		slice = nil
```

Comment:
I'm not sure it does what you think it does.

It makes the **function variable** `slice` point to nil. It doesn't change the underlying slice.

Reply:
It does not, thanks for pointing out. Resolved in [7b2f76b](url)